src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -835,72 +927,217 @@ private int decodeSession(
 
             return progress;
         }
-    }
 
-    private final class MqttServerStream
-    {
-        MqttServerStream()
-        {
-        }
 
-        private void onApplication(
-            int msgTypeId,
-            DirectBuffer buffer,
-            int index,
-            int length)
+        private final class Subscription
         {
-            switch (msgTypeId)
+            private final List<Byte> reasonCodes;
+            private final int ackCount;
+
+            private int successMask;
+            private int ackMask;
+
+            private Subscription(
+                int unrouteableMask,
+                int ackCount)
             {
-            case BeginFW.TYPE_ID:
-                final BeginFW begin = beginRO.wrap(buffer, index, index + length);
-                onBegin(begin);
-                break;
-            case EndFW.TYPE_ID:
-                final EndFW end = endRO.wrap(buffer, index, index + length);
-                onEnd(end);
-                break;
-            case ResetFW.TYPE_ID:
-                final ResetFW reset = resetRO.wrap(buffer, index, index + length);
-                onReset(reset);
-                break;
+                this.ackMask = unrouteableMask;
+                this.ackCount = ackCount;
+                this.successMask = 0;
+                this.reasonCodes = new ArrayList<>();
+                for (int i = 0; i < ackCount; i++)
+                {
+                    reasonCodes.add((byte) 0x8f);
+                }
```

Comment:
We don't need `reasonCodes` any more, since they are derivable from `successMask`, right?